### ISSUE TITLE
BitBucket Cloud: Git polling fails with "Invalid username or password" using OAuth credentials

### ISSUE THREAD

**Issue by yeroc** [jenkinsci/bitbucket-branch-source-plugin#328]
### Your checklist for this issue

- [x] Jenkins version 2.235.1

- [x] Plugin version 2.8.0

- [x] Bitbucket cloud

- [ ] Bitbucket server and version

### Description

I have a multi-branch project configured using OAuth credentials and the plugin is able to successfully scan for new branches and PR requests but on each discovered branch/PR the Git Polling Log is showing errors:

```
Started on Jul 15, 2020 9:29:00 PM
Using strategy: SpecificRevisionBuildChooser
[poll] Last Built Revision: Revision c00833ab7c4fb23999c19a37f52a633eca0266d0 (master)
using credential bitbucket-oauth-id
 > git --version # timeout=10
 > git ls-remote -h -- https://x-token-auth@bitbucket.org/orgname/base-java.git # timeout=10
ERROR: polling failed in /space/jenkins-agent/workspace/base-java_master on cal-dev1-agent
hudson.plugins.git.GitException: Command "git ls-remote -h -- https://x-token-auth@bitbucket.org/orgname/base-java.git" returned status code 128:
stdout: 
stderr: remote: Invalid username or password
fatal: Authentication failed for 'https://x-token-auth@bitbucket.org/orgname/base-java.git/'

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2436)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:2050)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1950)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1941)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.getHeadRev(CliGitAPIImpl.java:3311)
	at hudson.plugins.git.GitSCM.compareRemoteRevisionWithImpl(GitSCM.java:715)
	at hudson.plugins.git.GitSCM.compareRemoteRevisionWith(GitSCM.java:674)
Caused: java.io.IOException
	at hudson.plugins.git.GitSCM.compareRemoteRevisionWith(GitSCM.java:676)
	at org.jenkinsci.plugins.workflow.job.WorkflowJob.poll(WorkflowJob.java:615)
	at hudson.triggers.SCMTrigger$Runner.runPolling(SCMTrigger.java:605)
	at hudson.triggers.SCMTrigger$Runner.run(SCMTrigger.java:651)
	at hudson.util.SequentialExecutionQueue$QueueEntry.run(SequentialExecutionQueue.java:119)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Done. Took 1.1 sec
No changes
```

I've confirmed the BitBucket Cloud OAuth permissions are fine by manually getting an `access_token` using the same key & secret configured in Jenkins:
```
curl -X POST -u "key:secret" \
  https://bitbucket.org/site/oauth2/access_token \
  -d grant_type=client_credentials
```
and I'm able to run the same `git` command above successfully using the `access_token` so there's no issue with the permissions.

Here's the job-dsl config for the job:
```
multibranchPipelineJob('test-repository') {
  branchSources {
    branchSource {
      source {
        bitbucket {
          id('test-repository')
          credentialsId('bitbucket-oauth-id')
          repoOwner('orgname')
          repository('test-repository')
          traits {
            bitbucketBranchDiscovery {
              strategyId(1) // exclude branches which are also a PR
            }
            bitbucketPullRequestDiscovery {
              strategyId(2) // The current pull request revision
            }
          }
        }
      }
    }
  }

  orphanedItemStrategy {
    discardOldItems {
    }
  }

  triggers {
    periodicFolderTrigger {
      interval("10m")
    }
  }
}
```
And in the project `Jenkinsfile`:
```
pipeline {

  triggers {
    // check for updates every 15 minutes...
    pollSCM 'H/15 * * * *'
  }

...
}
```

**Comment by yeroc** [comment]
Examining the generated `config.xml` for the branch jobs it looks like the issue is that the plugin is generating a GitSCM configuration with `<userRemoteConfigs>` containing the temporary OAuth `access_token`.  Of course the OAuth token expires after one hour so soon after this stops working.  Here's a snippet of the generated config:
```
<?xml version='1.1' encoding='UTF-8'?>
<flow-definition plugin="workflow-job@2.39">
[...]
  <properties>
    <org.jenkinsci.plugins.workflow.multibranch.BranchJobProperty plugin="workflow-multibranch@2.21">
[...]
        <scm class="hudson.plugins.git.GitSCM" plugin="git@4.3.0">
          <configVersion>2</configVersion>
          <userRemoteConfigs>
            <hudson.plugins.git.UserRemoteConfig>
              <name>origin</name>
              <refspec>+refs/heads/PORTAL-1294:refs/remotes/origin/PR-82</refspec>
              <url>https://x-token-auth:{expired-token}@bitbucket.org/orgname/base-java.git</url>
              <credentialsId>bitbucket-oauth-id</credentialsId>
            </hudson.plugins.git.UserRemoteConfig>
          </userRemoteConfigs>
[...]
```
It's unclear to me whether this was an oversight or whether the these branch configurations are intended to be continuously regenerated?  If so, that is not happening.

I'm not familiar with Jenkins plugins.  Can anyone point me to where I might start looking?  It looks like the logic on how jobs are generated is at least partially located in `BitbucketGitSCMBuilder` but I'm not entirely sure...

**Comment by bitwiseman** [comment]
@yeroc  Interesting, good analysis.  Thanks! 

This is almost certainly an oversight / bug.   OAuth credentials were introduced in v2.8.0 (PR #262).

**Comment by yeroc** [comment]
This remains an issue as far as I can tell so believe it should be left open?

**Comment by nfalco79** [comment]
Is this issue still present with latest version? We have cloud instance with OAuth2 credential and works

**Comment by nfalco79** [comment]
Fixed by #926
